### ISSUE TITLE
Extended support for the non-native bit-order

### ISSUE THREAD

**Issue by ARG-Editor** [Ada-Rapporteur-Group/User-Community-Input#86]
This issue continues an unfinished topic from Ada 2022. This issue
was created to fulfill the ARG resolution of November 10, 2022.
--------------------------
Adapting code to run on targets with different endiannesses has long been a problematic endeavor for Ada users. Even though Ada 2005 improved the situation compared to Ada 95 with the AI95-0133-1 binding interpretation, there was in general no way to specify a record representation clause in a way that would produce identical sequences of bytes in memories for little-endian and
big-endian targets. In some cases, this could be achieved using customized representation clauses depending on the target, but in cases where some record components cross byte boundaries, there was no other working option but to
introduce explicit byte swapping operations.

A new attribute Scalar_Storage_Order was proposed to improve this situation. See AI12-0218-1 for details of the proposal. (http://www.ada-auth.org/cgi-bin/cvsweb.cgi/ai12s/ai12-0218-1.txt)

**Comment by ARG-Editor** [comment]
Here are my personal comments on this topic.

First, it's not clear to me why one would even want a non-native bit order in memory. In general, management of memory is best left to the compiler, in order to get the best possible performance (both in time and space) for the given target.


So the use of non-native orders should be limited to interfacing, most of which is associated with streaming. The proposal notes that it has no effect on streaming, so it does not help there. It also cannot help with hardware interfacing, as that necessarily is part of the current machine and thus uses the native bit-order.

The current facilities that Ada supports were defined in order to avoid requiring support for components that would be non-contiguous in the native bit order. Accessing such components would require so-called byte swapping (although the actual problem is more complex than just that). This would be substantially larger and slower code than needed to just access bit-fields (and that is already much larger and slower than native access to components).

Making it too easy to write such code could easily result in code that runs much slower than that defined in another language. This would generally be reflected back on Ada and on the Ada implementation, rather than the suboptimal code associated. It should be hard to write code that requires byte swapping!

Adding non-contiguous component support is pervasive in an Ada compiler; almost all access code would have to be able to support it. Doing so would be a fairly large project. But such non-native code should be quite rare in practice (as noted previously); the cost is much higher than the benefit (few programmers would benefit, and only in a small section of code).

Finally, there is no formal language proposed for this idea. Someone interested would need to step up and create a complete proposal including wording. In the absence of that, this proposal is not going anywhere.

**Comment by jeremiahbreeden** [comment]
I would greatly find this useful.  As an embedded programmer, in almost every project I do, I find myself having to use objects that need different endianess (usually due to interfacing with another IC to my process or a some sort of communications protocol that has a specific endianess but must interface with two separate processors of different endianess).  

I basically have to write either some bytes swapping routines or shift/or logic for these things instead, so having the compiler handle it when I specify it to would be nice.

On a more fun note, I recently started writing an emulator which targets a chip with a specific endianess, but I would like my emulator to be compilable on any platform that has a valid compiler, being able to set the endianess of some of the emulated components simplifies the code and makes it more readable than having to intersperse shift/ors (or byte swaps).

I don't find the slower speed argument that compelling to be honest.  The default byte order would be native, so the average programmer would not find it slower, and the folks that use this type of feature will most likely already be aware of the speed implications of mixing endianess, even via aspect

**Comment by CKWG** [comment]
Is there necessity for the attribute 'Scalar_Storage_Order? If I understand the proposal correctly, it stores objects, if the non-native bit order (NNBO) is defined, in a scrambled form and byte-swaps for each access. This looks like wasting processor load.

The need for the Ada 95/2005 'Bit_Order attribute is, as I see it, only present if data exchange between machines of different endianness is done. It is true that, when transferred, the data is acrambled and a byte swap is needed just once. After the swap, the data can be accessed in the normal way.
Let me explain my view in the way I used in the Ada Wikibook
https://en.wikibooks.org/wiki/Ada_Programming/Attributes/%27Bit_Order.

Let's take the example given in the AdaCore proposal
https://gcc.gnu.org/onlinedocs/gcc-4.9.4/gnat_rm/Attribute-Scalar_005fStorage_005fOrder.html
on a BE machine, i.e. in native order. This is how it looks:
```
for Date use record
  Years_Since_1980 at 0 range  0 ..  6;
  Month            at 0 range  7 .. 10;
  Day_Of_Month     at 0 range 11 .. 15;
end record;
for Date'Bit_Order use System.High_Order_First;

Byte 0  ->   1
BE   0123456789012345
     YYYYYYYMmmmDDDDD
```
We write consecutive bytes from left to right (as usual) and also count bits 0 to 15 (and further if needed) from left to right across byte boundaries. The LSB part of Month, which is on the next byte, is written in lower case.
When this is transferred byte by byte to a LE machine, this is the result:
```
Byte        1   <-  0
LE   5432109876543210
     mmmDDDDDYYYYYYYM
```
For convenience, we now write consecutive bytes from right to left and also count bits from right to left (again across byte boundaries). You see the data scrambled. Now byte swap:
```
Byte        1   <-  0
LE   5432109876543210
     mmmDDDDDYYYYYYYM  after transfer as above
     YYYYYYYMmmmDDDDD  after byte swap
NNBO 0123456789012345  count within machine scalar left to right
```
You see that the attribute Date'Bit_Order in the NNBO correctly describes the outlay. This has been the current state since Ada 2005.

As I understand the proposed attribute 'Scalar_Storage_Order, it will force storage in non-native order (on a LE machine in our example) in the scrambled form and implicitly byte-swap on each access so that the data may be shifted freely between machines without the explicit byte-swap that is required without the new attribute.

This direly looks like a waste of processor load!

So I see two model use cases:

1.) I'm thinking of data created on one machine, transferred to another one and processed there, at the end transferred back. Thus, byte swapping is needed exactly twice in this process, not for each access. Hence the current Ada 2005 solution is sufficient (writing a byte swap is easy).

2.) In distributed systems, which find more and more applications, perhaps the data lies in some separate neverland between machines (and not in two copies in the memory of each one as in the first use case) - on a shared passive library unit perhaps (I have no experience with these), where the data is directly accessed without transfer. Here, this attribute makes sense.

I have no idea how often these two use cases are.

**Comment by sttaft** [comment]
The fact is that the Scalar_Storage_Order aspect is used pretty heavily in the GNAT user community, so apparently the use cases you identify (or other use cases that exist) come up quite often.  That seems to be the best argument for it.  Many GNAT users make use of it, and so it is definitely something that at least those users would want to be portable across compilers.

**Comment by CKWG** [comment]
So is my analysis correct and byte swapping is done automatically?
As I understand, you would classify my reluctance wrt Scalar_Storage_Order as premature optimization - just let the compiler, ahem the machine, do the work and only measure if there are bottlenecks.
If so, so be it. User communiity vote. (That's what this whole github thing is about.)
I would then propose:

Just use the Scalar_Storage_Order alone, the Bit_Order being implied consistent. (If both are given, they must of course use the same value.)

**Comment by sttaft** [comment]
> I would then propose:
> 
> Just use the Scalar_Storage_Order alone, the Bit_Order being implied consistent. (If both are given, they must of course use the same value.)

I agree this would be simpler in some cases, but in general in Ada we aim for being pretty explicit, and it is easy enough to specify both when specifying either.  By being explicit, we know that both the programmer and the reader/maintainer understand what are the implications.

**Comment by CKWG** [comment]
> Many GNAT users make use of it, and so it is definitely something that at least those users would want to be portable across compilers.

Hm, this is about Ada 2022, most probably this feature will not be in Ada 2022 TC1, which I understand will hold only corrections, not new features. This directly leads me to the question about the portability of the current standard:

Skimming through the list of current compilers, methinks the complete implementation of Ada 2012 (core) is GNAT alone, Ada 2022 is only partly implemented as yet (esp. **parallel** keyword). Other compilers still only offer Ada 95 or 2005. Is there a standstill?

So I dare a heretical statement - is there any portability left?

The community, small as it may be, raises more and more wishes and the poor compiler writers cannot or are reluctant to follow.

On the other side, hardware evolves and languages strive to offer features for handling it (see multithreads e.g.).

Is there a solution for this dilemma? Can there even be one?

**Comment by sttaft** [comment]
> Skimming through the list of current compilers, methinks the complete implementation of Ada 2012 (core) is GNAT alone, Ada 2022 is only partly implemented as yet (esp. parallel keyword). Other compilers still only offer Ada 95 or 2005. Is there a standstill?

PTC has implemented Ada 2012 (https://www.ptc.com/en/products/developer-tools/objectada#announcements).  Not sure what plans they have for Ada 2022.

RR Software's compiler also has implemented parts of Ada 2012 (http://www.rrsoftware.com/html/blog/ja-321a-rel.html).

There is a move afoot to implement more parts of Ada 2022 in the FSF GNAT.

There is a version of the ParaSail interpreter that accepts Ada 2022 syntax (https://github.com/parasail-lang/parasail/tree/main/ada202x_parser), and has a lot of support for the parallel features (no surprise there! ;-).

So portability is still relevant, and is kind of the point of having an International Standard.

Note that Ada is not alone in having various compilers providing support at different levels of the corresponding standard ...

**Comment by ARG-Editor** [comment]
Tucker says:


> The fact is that the Scalar_Storage_Order aspect is used pretty heavily
> in the GNAT user community, so apparently the use cases you identify
> (or other use cases that exist) come up quite often. That seems to be
> the best argument for it.  

	I still would like to find out what use cases it is actually used
for in the GNAT user community. Perhaps there are some compelling ones, but
most of the ones proferred so far are dangerous in that they make Ada code
have horrible performance compared to other languages.

	There is a language design principle that expensive constructs
should look expensive, and Scalar_Storage_Order violates that principle
pretty significantly. Using Unchecked_Conversion to do the byte-swapping
looks appropriately expensive (probably more so that it actually is,
although GNAT used to [might still, I haven't checked] have a bad
implementation of IC which made unnecessary copies of the objects). So it is
preferred from a purist perspective.

	As such, I wanted to see some actual use cases in order to convince
myself that the aspect was worthwhile compared to just properly doing the
needed Unchecked_Conversions.

> Many GNAT users make use of it, and so it is definitely something that at
> least those users would want to be portable across compilers. 

	I seriously doubt that many if any other Ada compilers would support
this aspect even if it was in the Ada Standard. (Remember that all vendors
are daisy picking features they like these days, no one is implementing
everything.) The cost-benefit analysis on this one is heavily skewed to the
cost side. While a lousy implementation might cost less, anything that
attempts to use existing machine support would require a lot of back-end
support, and adding new support to a back-end is very disruptive (as almost
all of the tools need modification). I'd guess that we would be talking at
least 3-6 man-months of work for Janus/Ada. There are many other pending
projects for Janus/Ada that have much more general applicability that will
take similar amounts of effort.

     And that effort would be for something that shouldn't occur very often,
as security advice is to avoid binary protocols completely (using text
communication instead, which avoids this problem). For instance, while the
Internet is technically big-endian, most protocols are purely text and thus
there is no real impact when using the Internet from little-endian machines.
(Which again is why I would like to hear more about the real-world use
cases, as opposed to whatever us ARG members might imagine the uses are.)

	                Randy.

**Comment by Jeff-Cousins** [comment]
*
Note that Ada is not alone in having various compilers providing support at different levels of the corresponding standard

The editor of the last revision of the Fortran standard told me that no vendor has implemented the full standard for donkey's years.

**Comment by CKWG** [comment]
> Note that Ada is not alone in having various compilers providing support at different levels of the corresponding standard
Remember that all vendors are daisy picking features they like these days, no one is implementing everything.

This definitely means: There will be no more validated Ada compilers.

> they make Ada code have horrible performance compared to other languages.

What I don't understand: Byte swapping seems cheap compared to mask and shift operations for accessing components.

**Comment by ARG-Editor** [comment]
>> they make Ada code have horrible performance compared to other
languages. 
 > What I don't understand: Byte swapping seems cheap compared to mask and
shift operations for accessing components. 
 
How so???  

On most machines (and certainly the Intel machines that I am used to), mask
and shift is two in-register instructions for reading, and one extra read
(sometimes) and four in-register instructions for writing.

Moreover, one usually can ignore the non-memory instructions for performance
reasons (a rough estimate of performance can be made by just adding up all
of the reads and stores; they're far more expensive than the other
operations). So there is almost no effect on reads and just a doubling (at
worst) on stores (which are much less common in most code than reads).

OTOH, most machines don't have byte-swapping instructions, so one ends up
with a lot of extra reads and writes (4 times plus at least 4 shifts for
reading) OR one has to use extra registers (which in the case of the Intel
processors with very few registers means other things can't be in registers,
increasing the number of reads and writes for them).

This means that a byte swapping reading for a 32-bit item can be 4 times
slower and 8 times larger than a native read. (Larger code tends to be
slower as well.)

And if your backend doesn't support byte swapping directly, then you have to
emulate it much as an Ada programmer would -- and that means the code would
be even worse.

Finally, I don't recommend using unaligned or packed components for much,
either, as they are substantially more expensive (especially writes) than
native sized and aligned components (the layout you get by default with most
Ada compilers). They should be strictly limited to interfacing or long-term
storage of entities, not anything that gets used a lot. This is another
thing that is very commonly abused (and can make Ada look very bad compared
to some other language).

                     Randy.

**Comment by ARG-Editor** [comment]
> This definitely means: There will be no more validated Ada compilers.

There haven't been any since 2001 (those that passed formal testing).

And my experiments suggest that it is unlikely that any existing compiler
would pass formal testing without significant fixes. So this has been true
for a long time. There may be a difference in degree in the future, but I
suspect it is less of a difference than most understand.

[Sorry, I can't give any details; vendors give me access to their technology
to use for ACATS development work with the understanding that I won't
publicize what works and does not work. That's in part because the versions
I have access to can be close to the cutting edge and thus may have issues
yet to be fixed.]

                 Randy.

**Comment by ARG-Editor** [comment]
>> What I don't understand: Byte swapping seems cheap 
>> compared to mask and shift operations for accessing components. 

> How so??? 

A more abstract way to look at this is to remember that when you access a
bit-mapped component, that is a single piece of data to manipulate. Storing
such a component means manipulating two pieces of data (the original value
for the composite and the new value for the component).

But byte-swapping a 32-bit integer means that you have 4 pieces of data to
manipulate; and for a 64-bit float you have 8 pieces of data to manipulate.
(Each byte has to go to a different new location, there's no way to combine
the operations.) It's hard to see how you could do that more simply than
handling one or two pieces. (Assuming no direct hardware support.)

For floating point types (which have their own registers separate from the
integer registers), you pretty much have to write them back to scratch
memory (there's no integer <=> float register operations on Intel machines).
And adding reads/writes is the worst performance thief that you have.

                Randy.

**Comment by ArnaudCharlet** [comment]
﻿
> > This definitely means: There will be no more validated Ada compilers.
> 
> There haven't been any since 2001 (those that passed formal testing).
Right.

> 
> And my experiments suggest that it is unlikely that any existing compiler
> would pass formal testing without significant fixes.

For the record since this discussion is tracked, the above statement is wrong for gnat pro up to Ada 2012 where we internally validate our compiler everyday and produce self certificates for our customers.

Arno

**Comment by ARG-Editor** [comment]
Arnaud Charlet writes:
	
>> And my experiments suggest that it is unlikely that any existing compiler

>> would pass formal testing without significant fixes. 
	
>For the record since this discussion is tracked, the above statement is
wrong
>for gnat pro up to Ada 2012 where we internally validate our compiler
everyday
>and produce self certificates for our customers. 

Formal testing is a very different beast than self-testing, no matter how
well-intentioned and meticulous it is.

In formal testing, an independent tester is required to grade all of the
ACATS results for a run of the current ACATS test suite without any
reference to previous results or vendor information.

There then is a phase where the tester and the implementor come to an
agreement on all of the tests with possibly non-conforming results (based on
the tester's grading). Given that test grading of B and L tests is rather
subjective, this is typically a fairly long list. The implementor has to:
  (A) convince the tester that the results are actually conforming;
  (B) convince the tester that different processing of the tests produces
      conforming results;
  (C) dispute the test as incorrect with the ACAA;
  (D) appeal the results to the ACAA (and possibly beyond);
or(E) fix the compiler to give results that are acceptable to the tester.

Most of the issues would likely be resolved with one of the first two
options. But a few would remain.
Given the fact that disputes and appeals have uncertain timelines and
uncertain probabilities of success, and given business needs to finish
formal testing (if one starts it in the first place), it is highly likely
that some fixes to the compiler are better choices than the alternatives.
Even if none are "wrong" in the eyes of the vendor; the tester has the last
say in the case of formal testing. That's the primary basis for my
statement.

Another problem with self-testing is the reliance on some form of "known
good results" to determine the grading of B-Tests and L-Tests. These "known
good results" need to be modified periodically to deal with changes in error
handling and test fixes/improvements. Over time, errors creep in. When
testing the ACATS grading tools on Janus/Ada, I was surprised to find nearly
a dozen instances where the "known good results" were not good at all. I can
only put that down to 20+ years of modifications adding occasional mistakes.
Only independent testing can truly provide/verify "known good results"
(although the ACATS grading tools can reduce the number of tests with
subjective results to less than 10%, which helps mitigate the problem).

Without actually going through the formal testing process, it is unknowable
how any particular implementation would fare. That's one reason why I
haven't shared the details of my grading experiment; the vendor would have
to have an opportunity to explain any nonconforming results. And I was not
trying to determine the quality of any implementation; I was doing the
exercise to validate the ACATS grading tools. So going further was not
necessary.

But I still doubt that any existing implementation would pass formal testing
without fixes. For any ACATS version. Self-testing proves nothing; it is
merely a starting point for formal testing. Simply because it as been so
long since any formal testing has been done on any compiler that there would
be a long list of discussion items, and some of those would end up resulting
in compiler fixes. That really doesn't say anything about the quality of any
particular implementation, but more the subjectiveness of grading the ACATS
(something I wanted to change, but did not have enough support to actually
accomplish).

               Randy.

**Comment by CKWG** [comment]
Thank you, Randy, for these clear words. They reflect my feelings when I read Arnaud's post. It's a pity and a shame that there are no more formal testings - let alone certificates.

I see Ada approximating the path all other languages take. They dissolve into dialects or - worse - don't even have proper reference manuals... The claim that Ada initially had and maintained for a long time is gone.

Do we as users have a share in this development? I see many questions in forums that could be answered with a glance at the RM.

**Comment by ArnaudCharlet** [comment]
>
> >> And my experiments suggest that it is unlikely that any existing
> compiler
> >> would pass formal testing without significant fixes.
>
> >For the record since this discussion is tracked, the above statement is
> wrong
> >for gnat pro up to Ada 2012 where we internally validate our compiler
> everyday and produce self certificates for our customers.
>
> Formal testing is a very different beast than self-testing, no matter how
> well-intentioned and meticulous it is.
>

We at AdaCore know very well what formal testing entails, we've done it a
number of times in the past and we follow a rigorous QA and review process
when e.g. changes such as baseline updates need to be made (which doesn't
occur that often in practice for GNAT).

Your original statement which I replied to was: *my experiments suggest
that it is unlikely that any existing compiler **would pass formal testing
without significant fixes.*

The above statement is simply wrong for GNAT Pro: AdaCore would be in a
position to formally validate GNAT Pro today if there was a business need
for it, without having to make "significant fixes" as you claim. You're
free to "doubt it" with the (partial) knowledge that you have, as long as
it's clear to everyone that this is your own personal judgment and as long
as you do not present it as truth or as something "very likely", in
particular when you get a statement from a compiler vendor that this isn't
the case.

**Comment by ebotcazou** [comment]
The main impetus for the design and implementation of Scalar_Storage_Order at AdaCore was that engineers were fed up with the constant flow of questions and remarks about Bit_Order, arguably one of the most confusing features of the language.  Explaining to the average user what the compiler does with a bare Bit_Order set to System.High_Order_First on a little-endian system is tricky and, once understood, inevitably led to the question (at the time): OK, but how can I have the same memory representation as on my PowerPC board?  Scalar_Storage_Order very likely saved many hours of (frustrating) customer support.

**Comment by ebotcazou** [comment]
The argument based on the language design principle that expensive constructs should look expensive (and Scalar_Storage_Order violates that principle) is a valid one, but the problem is that this principle has already been violated so many times in the language that one can no longer reasonably apply it.

**Comment by ebotcazou** [comment]
IMO the main argument against Scalar_Storage_Order is that its implementation is no small undertaking, especially in an existing production compiler.  It was made possible in GCC/GNAT with reasonable run time performances only under the following assumption: every scalar object (in memory) is accessed during the entire execution of the program with a single storage order (either High_Order_First or Low_Order_First); if this prerequisite is violated by the program, then its execution is erroneous.  This rules out fiddling with the storage order by means of aliasing, type punning, etc; only an explicit Unchecked_Conversion is allowed to toggle the storage order.

**Comment by sttaft** [comment]
Thanks, Eric, for the background and information about the current implementation.

**Comment by ARG-Editor** [comment]
> The argument based on the language design principle that expensive
> constructs should look expensive (and Scalar_Storage_Order violates
> that principle) is a valid one, but the problem is that this
> principle has already been violated so many times in the language
> that one can no longer reasonably apply it.

I disagree with everything after "but" here. I realize that what looks
expensive is somewhat in the eye of the beholder, but I disagree that there
are many violations in current Ada. The vast majority of them are either
caused by dubious usage (overloading "+" with an very expensive routine, for
instance), or by avoidable design decisions (:= can be expensive, but only
for non-contiguous objects, which most compilers don't use). In particular,
"." is always cheap when accessing components (indeed, it can only be
expensive if it is hiding a parameterless call -- that was impossible in Ada
80 but the call marker was removed in Ada 83. Something I really hated at
the time, but no one was asking a random grad student at the University of
Wisconsin what he thought. :-) I strongly dislike making it 10 times more
expensive an hiding it.

But regardless of that, I don't think it is particularly relevant what the
language already does. The idea that the language violates some useful
property in some case, so we shouldn't worry about that property in any
case, guarantees a race to the bottom. We would end up turning Ada into a
mess, like <insert name of popular other language here>. It is likely you
can say that about any of the usual Ada properties (readability,
orthogonality, and so on), but that doesn't mean that we shouldn't try to
preserve them.

                     Randy.

**Comment by ARG-Editor** [comment]
Surely, Bit_Order as defined was a mistake. We should have had the guts to
say "No" when pressured to support some sort of non-native bit order. Since
we did not have the desire to make something insanely expensive, we ended up
with a useless watered-down feature. Bit_Order is one of those things that I
have no plans to ever implement in Janus/Ada, simply because it is actively
harmful for the reasons you give. No reason to sign up for that.
 
                           Randy.


  _____

**Comment by ARG-Editor** [comment]
>IMO the main argument against Scalar_Storage_Order is that its
>implementation is no small undertaking, especially in an existing
>production compiler. It was made possible in GCC/GNAT with
>reasonable run time performances only under the following
>assumption: every object (in memory) is accessed during the
>entire execution of the program with a single storage order
>(either High_Order_First or Low_Order_First); if this
>prerequisite is violated by the program, then its execution
>is erroneous. This rules out fiddling with the storage order
>by means of aliasing, type punning, etc; only an explicit
>Unchecked_Conversion is allowed to toggle the storage order. 

Interesting. That seems to imply that you *don't* want the aspect
standardized, since one could have no such assumption in standard Ada code.
Certainly, the ACATS tests would use both byte orders and certainly would
not use Unchecked_Conversion to go between them. (The entire point of having
such an aspect IMHO seems to be to avoid using Unchecked_Conversion; if
you're willing to do that, you need no aspect at all.) It sounds like
supporting such tests would be a significant job for GNAT as well.

While the restriction you describe makes much more sense than the general
aspect that works anywhere, it also means that (in my view at least) the
aspect is not a candidate for standardization. It seems more like a hack
given the restriction. (Nothing wrong with hacks, getting the job done is
important, but they don't belong in the RM.)

Thanks for explaining that this isn't implemented generally in GNAT.

             Randy.

P.S. I think this topic should be closed as no action based on this new
information, but as I've always opposed this aspect I definitely don't think
I should be making this call.

**Comment by joshua-c-fletcher** [comment]
The handling of endianness for portability is actually something I really like in Ada.

Incidentally, before 'Bit_Order in Ada 2005+, we had the following paper expressing a way to describe records for portability, when the records violate the byte boundaries:
https://www.sigada.org/ada_letters/sept2005/Endian-Independent%20Paper.pdf

Such records would essentially have their elements laid out in reverse, depending on the endianness of the host, and require "Wholesale Bye Reversal".  'Bit_Order in Ada 2005+ works exactly the same way as the fancy clauses in that paper... which is perfect, because other native elements need Byte Reversal, too, when converting to an external representation with the opposite endianness of the host.

In Ada, Streams are a language-native way to distinguish between a local representation and an external representation.  If you have a Stream Type that 'knows' whether the external representation should be High_Order_First or Low_Order_First, and if the native byte order of your host is known at compile time (typically the same as System.Default_Bit_Order), then your stream type can know whether to conditionally swap the stream elements in the Stream_Element_Array 'item', when invoked.

This assumes records and arrays invoke the streaming procedures for each element in turn, except where the streaming subprograms are overridden, which is consistent with how it is described in the RM.  For GNAT, you'd need to use the No_Stream_Optimizatioins pragma, because they stream strings as a single 'item' by default, rather than like a typical array.

For objects without any custom layouts via representation clauses, if your stream type conditionally does wholesale Stream_Element reversal of each item passed to the streaming functions, the external representations will match even if the local representations are consistent with hosts of different endianness.

If the external representation needs to include elements that don't respect byte boundaries, the method described in the Andress paper can be used (without 'Bit_Order) or 'Bit_Order can be used to lay out the record equivalently (and more simply).

The streaming functions for such records could unchecked_covert or overlay the record with a Stream_Element_Array of the same size and call the streaming function with the stream_element_array item.  if the Stream type being used conditionally reverses the stream elements based on the selected endianness of the stream (representing the external representation) versus the host, then again: it just works.  The external representation will match, even while the local representations are consistent with hosts of different endianness with respect to byte order.

Since 'Bit_Order doesn't do any re-ordering of bytes, and simply causes the numbers in the representation clause to be interpreted from the other end, it ('Bit_Order, that is) shouldn't result in _any_ extra overhead at runtime.  Streaming and conditional swapping of elements adds some overhead, but some optimizations are possible (even in the stream type itself).  It appears that extra overhead exists in some form anyway, even if using 'Scalar_Storage_Order, which isn't needed in this solution.

Note: It isn't always necessary to rely on Streaming to do the conditional wholesale reversal.
If the interface in question is word-based, rather than byte-based, the driver for such an interface might already byte-swap the "words" to native order anyway, in which case you just make sure your records match the word size and set 'Bit_Order on them.  The driver does the wholesale reversal and if such records are defined with 'Bit_Order, it just works.

I don't mind 'Scalar_Storage_Order existing or being added to the language, but I feel that the existence of Ada.Streams, System.Default_Bit_Order, and the 'Bit_Order attribute already allow Ada to be used to produce portable, endian independent code with low development overhead; the stream type can do most of the work.  I see it as one of Ada's strengths.

**Comment by ebotcazou** [comment]
My wording was either imprecise or your understanding of it slightly wrong, but Scalar_Storage_Order is implemented quite generally in GNAT under the restriction I mentioned (every scalar object is accessed using a single storage order during the entire execution of the program), which is an extremely weak restriction; it says that you must (morally) partition your memory space into the big-endian part and the little-endian part.  Once this is done, everything works transparently.  In almost all cases, the only thing to do is to add the Scalar_Storage_Order attribute on the types without any other code change.  The compiler will try to detect the common cases where the restriction is violated (typically overlays toggling the storage order) and warn about them.

I don't have a strong opinion as to whether this should be standardized or not.

**Comment by ARG-Editor** [comment]
Thanks for the explanation. To me, the statements "Scalar_Storage_Order is
implemented quite generally" and "the restriction I mentioned" are pretty
much mutually exclusive. A restriction on the use of the aspect seems to
violate its purpose. So further explanation is helpful.
 
It seems that the "proper" use of Scalar_Storage_Order requires at some
point both orders in the same code (how else can one create a native order
item to use in the rest of the program?). The partitioning seems to make
that difficult.
 
I also have to wonder what a "memory space" in the non-native order means on
most targets (which only support native order operations in hardware). For
instance, on an Intel IA32 target, there are no instructions for accessing
32-bit big-endian objects. Emulation is expensive (I'd rather support it
only when absolutely necessary). Moreover, well-structured code is likely to
only have big-endian values in a few subprograms. You'd never want to store
anything in main memory that way. (My biggest objection to this proposal has
always been that it is too easy to abuse; it seems like this partitioning
requirement forces some level of abuse.)
 
Anyway, if we standardize this aspect, it would be expected to work in usual
use-cases without any user-visible restrictions. Not knowing what actually
constitutes a "memory space", I can't tell if that would be a problem.
 
An ACATS test would probably simulate reading non-native records from a
stream, with a header record (containing a length of data records) and some
number of data records. The header record would be read and some math would
be done on the length to check that is is sensible (probably would not
convert to the native order for that), then the data records would be read
in and converted into a native order buffer. That would look something like:
 
     type Native_Data is record ...
     type Non_Native_Data is new Native_Data;
     for Non_Native_Data use record
         ... -- a definition using Scalar_Storage_Order for the components.
     end record Non_Native_Data;
 
     for I in 1 .. Header.Length loop
         Buffer (I) := Native_Data (Get_Non_Native_Data_from_Stream);
     end loop;
 
Get_Non_Native_from_Stream being a function that returns Non_Native_Data.
Perhaps it would make more sense to put the type conversion inside the
function call and return a value of Native_Data. Both really ought to work,
as we don't want to be dictating designs on users.
 
                     Randy.
 


  _____

**Comment by jere-software** [comment]
I know on my end I run into this problem very often.  Every time we start a new project we end up interfacing a LE system with either a BE system or BE message protocol or BE hardware interface.  One of the things I love about Ada is that it tries to take the more dangerous or complicated implementation work away from the programmer so they can focus on the design.   This seems like a no brainer to me.   It should be an available language defined aspect.  If a compiler vendor doesn't want to implement it, then they don't and they document it (like any other Ada feature).  The feature enhances both maintainability and readability.  Also having Bit_Order on its own is just a trap for the programmer.

**Comment by CKWG** [comment]
i Jere,

which of my two model cases above do you encounter?

> So I see two model use cases:
> 
> 1.) I'm thinking of data created on one machine, transferred to another one and processed there, at the end transferred back. Thus, byte swapping is needed exactly twice in this process, not for each access. Hence the current Ada 2005 solution is sufficient (writing a byte swap is easy).
> 
> 2.) In distributed systems, which find more and more applications, perhaps the data lies in some separate neverland between machines (and not in two copies in the memory of each one as in the first use case) - on a shared passive library unit perhaps (I have no experience with these), where the data is directly accessed without transfer. Here, this attribute makes sense.

I don't really understand what Eric Botcazou explains about the Gnat implementation, but have the impression it's number one.

**Comment by jere-software** [comment]
I have mostly `#1` with a bit of `#3` (not listed).   I have to generate my own byte swapping.  I get that we can write our own byte swapping operations.  We can do a lot of things manually in Ada, but it is error prone, hinders maintenance, and hinders readability.  I do believe it is a good idea to have an aspect that lets the compiler handle that for you, increasing portability and readability and limiting programmer made mistakes.

For `#3`, I am emulating a LE system and want the code to be compilable on any system that has a supporting Ada compiler (Big Endian, Little Endian, Middle Endian, custom).  Like `#1` I generate my own byte swapping routines, but again, I think the compiler can do a better job here.

**Comment by ARG-Editor** [comment]
> For #3, I am emulating a LE system and want the code to be
> compilable on any system that has a supporting Ada compiler
> (Big Endian, Little Endian, Middle Endian, custom). Like #1
> I generate my own byte swapping routines, but again, I think
> the compiler can do a better job here.

This is the holy grail, of course, and that was the original goal with
Bit_Order, but it is impossible. Any non-native bit order needs information
usually left to the compiler. In particular, you need to know the word size
for which a component will be accessed.

To deal with this, Bit_Order adds a lot of additional assumptions based on
how a layout is written. That means it takes extreme care to write a rep
spec for the non-native bit order.

Consider the following types:
    type Status_Type is (Off, On, Error);
    type Device_Reg_Type is record
       Data : Character;
       Status : Status_Type;
       Active : Boolean;
    end record;

You might write the rep spec for this type in either of two ways:
   for Device_Reg_Type use record
      Data at 0 range 0 .. 7;
      Status at 1 range 0 .. 1;
      Active at 1 range 2 .. 2;
   end record;
or
   for Device_Reg_Type use record
      Data at 0 range 0 .. 7;
      Status at 0 range 8 .. 9;
      Active at 0 range 10 .. 10;
   end record;

You'll probably chose one based on the register definition in a databook (or
from a type definition in some other language), and you'll probably just
copy that other definition.

For the native order, these two mean exactly the same thing. But for the
non-native bit order, these two rep specs have a completely different
meaning: the first implies no byte swap, and the second does.

There's probably little hint in the databook or type definition which of
these is correct. For Christoph's case #1, you probably would experiment
and/or use a hex editor to figure out what is correct. That's OK because it
is pretty limited. But that's not true for #3.

Moreover, for #3, I'd guess that most types wouldn't even have
representation clauses. In that case, what the non-native bit order does is
completely up the compiler. It doesn't even have to make the same decision
for the same definitions in different contexts. So while the code might run,
whether it is at all meaningful when run in the non-native order is
impossible to say.

In summary, this problem is not one that can be handled wholly by the
compiler. Even with the proposed aspect, the exact form of the
representation clauses has a massive effect on what is generated, and thus
their definition requires extreme care and information that might not even
be available. There is no magic bullet possible here.

                    Randy.

**Comment by CKWG** [comment]
Machine Scalar is defined in 13.3(8.1/3)
A machine scalar is an amount of storage that can be conveniently and efficiently loaded, stored, or operated upon by the hardware. Machine scalars consist of an integral number of storage elements. The set of machine scalars is implementation defined, but includes at least the storage element and the word. Machine scalars are used to interpret component_clauses when the nondefault bit ordering applies.

Gnat supports only 8, 16, 32 ... bits machine scalars. Thus a record like
```
  for Data use record
    A at 0 range  0 .. 12;
    B at 0 range 13 .. 17;
    C at 0 range 18 .. 23;
  end record;
  for Data'Bit_Order use System.High_Order_First;
```
cannot use the Bit_Order attribute on LE since Gnat requests a 32 bit machine scalar, whereas on the native BE only 24 bits are needed.

Is this correct or am I in error?

With the constants defined in the Randall Andress  paper, the rep spec reads (the values for First_Bit and Last_Bit to be used at the LHS are those for BE):
```
for Data use record                                                       --    LE        BE
  A at 0 range MSB_24 +  0 * F1 + 12 * F2 .. MSB_24 + 12 * F1 +  0 * F2;  -- 11 .. 23   0 .. 12
  B at 0 range MSB_24 + 13 * F1 + 17 * F2 .. MSB_24 + 17 * F1 + 13 * F2;  --  6 .. 10  13 .. 17
  C at 0 range MSB_24 + 18 * F1 + 23 * F2 .. MSB_24 + 23 * F1 + 18 * F2;  --  0 ..  5  18 .. 23
end record;
```

**Comment by joshua-c-fletcher** [comment]
Christoph, your observation appears correct in GNAT, and I find it troubling, because I've (relatively recently) seen 'Bit_Order as a good alternative equivalent to using the parameterized representation clauses described in the Andress paper I referenced earlier.

When Stream_Element_Size is byte-sized (not guaranteed by the RM, but true more than one compiler I've used), Streaming with a Stream type that conditionally reverses the order of elements in the Stream_Element_Array does a really good job of solving the byte-order problem, especially when using default 'Read 'Write behaviour for types, relying on the behaviour defined in the RM (streaming elements in canonical order, and predictable sizes for elementary types)

When records like your example are needed, where elements cross a byte boundary and start or end within a byte, I've typically used, for years, the parameterized representation clauses described in the Andress paper.

It turns out, if you look at the AI and discussion for "fixing" 'Bit_Order, Robert Dewar described the approach in Randall Andress' paper and says they worked together on it: http://www.ada-auth.org/cgi-bin/cvsweb.cgi/ais/ai-00133.txt?rev=1.17.  That remark was dated before the paper was published, but it is the same technique: https://www.sigada.org/ada_letters/sept2005/Endian-Independent%20Paper.pdf (notably, Dewar is mentioned, but not credited in the published paper)

The proposal for fixing 'Bit_Order (in AI95-00133/06) is described as being based on an earlier paper by Norman Cohen, of which Randall Andress' paper describes itself as a successor.  The link to that paper in the AI no longer works, but it can be found here: https://dl.acm.org/doi/pdf/10.1145/181492.181493

What I find deeply ironic is that, while Robert Dewar credits himself as helping to come up with the method in Randall Andress' paper, he nonetheless argued against adopting the 'Bit_Order proposal... even though the (fixed) 'Bit_Order attribute basically does exactly what his parameterized representation clauses do.  The difference is that (if you can accept the fact the 'Bit_Order does _only_ bit order and _not_ byte order, to ensure the parts of elements remain continuous across byte boundaries), 'Bit_Order is more readable than the representation clauses in his technique.  If not for the limitation you've pointed out, it makes the technique obsolete (or at least unnecessary)... because it does the same thing.

The fact is: those parameterized representation clauses work very well when they're needed, but they're ugly to look at and confusing to the uninitiated.  Setting 'Bit_Order isn't ugly, but it's certainly less verbose.  Even setting 'Bit_Order It can be confusing if you don't know how it works, but if you understand that it just orders the elements from the other end, it's not so bad, and much nicer to use, if it works.

Years ago, before GNAT implemented the AI95-0133-1 binding interpretation, I looked into using 'Bit_Order, found it didn't work beyond individual bytes.  I concluded that it was practically useless, and avoided it afterwards, writing it off.  Past that investigation, I continued using the parameterized representation clauses - which do work, and work well.  They're only needed in certain cases, but those cases do come up, especially for ICDs.  It was only relatively recently that I learned about the binding interpretation that allows 'Bit_Order to be more useful.  I tried it out for a number of cases that didn't work in the past, and indeed - for the cases that GNAT allows to compile, it does work as a suitable alternative to the parameterized representation clauses, and was no longer as limiting (or as useless) as the original '95 version.  It doesn't do byte order, but I'm happy about that - I have a solution for Byte Order I'm happy with which works with the language (as long as Stream_Elements are byte-sized), and my general solution works well with 'Bit_Order if it works like it does - minus the issue you've pointed out.  I was quite excited about 'Bit_Order being a way to reverse elements in a record based on the assigned record's bit order differing from the hosts - exactly what the parameterized clauses do.

... but the way I use it, I'd want it to work with 3-byte records, such as your example, Christoph.  As you noted, the GNAT implementation that doesn't actually work for that.  Therefore, I could replace many instances where I've used parameterized representation clauses with setting 'Bit_Order, but not those that conflict with the restriction you called out.  It seems like a rather arbitrary restriction.  There may be some rationale for it, or it may be related to some particular wording, but the way I'd want and expect it to work wouldn't need that limitation.

**Comment by joshua-c-fletcher** [comment]
Regarding Randy's comment about different rep clauses, where that 'at' byte either zero or non-zero
```
   for Device_Reg_Type use record
      Data   at 0 range 0 .. 7;
      Status at 1 range 0 .. 1;
      Active at 1 range 2 .. 2;
   end record;
```
vs
```
   for Device_Reg_Type use record
      Data   at 0 range 0 .. 7;
      Status at 0 range 8 .. 9;
      Active at 0 range 10 .. 10;
   end record;
```
I did experiment with scenarios like that (perhaps not enough) when newly considering 'Bit_Order after dismissing it for so long, and was pleasantly surprised to see it work far better than I'd expected, given my earlier experience with it.

The Andress paper always uses 'at 0' and 'wholesale reversal' for a whole record, but if you have an interface that actually provides everything in 16 or 32-bit words that all need to be swapped, you can handle that with incrementing the 'at' byte by 2 or 4, for each part, respectively.  In that case, you could use an 'MSB' constant (re: the paper) for that word size (smaller than the record), but have a larger record that is a multiple of those words (rather than treating the whole record as a giant word).  Instead of whole-sale reversal of the whole record, you'd swap each (shorter) N-Byte word in the record (equivalent to streaming an array of N-Byte unsigned elements with byte reversal for each).  That technique can be generalized pretty well, too... although it's pretty similar to writing the parts as separate records with rep clauses (with at 0, like in the paper) and then making a bigger record comprised of the smaller ones)

With the parameterized rep specs, the beginning and ending of each part getting reversed is mathematically definitive (whether you do it all as one giant "word" with "at 0" every time, or whether you make records of several words with elements separately reversed in each)

With 'Bit_Order it is admittedly _not_ obvious where the 'other' end is: the 'at' part identifies the start, but nothing clearly shows where the end of that part is for the non-native order.  If there are no gaps, though, there'd only be one way to fit in the elements, and it seemed to work as I wanted it to when I tried it out experimentally.

I didn't realize the successful experiments were due to compiler-specific interpretations, and concluded that 'Bit_Order finally had reliable behaviour.

I did consider this ambiguity, though; for example:
```
   Something at 0 range 0..0;
```
On a non-native bit order, the behaviour I hoped for could interpret that line differently depending on the other content in the record.
It would be equivalent to local 7..7 if content for the next group of elements starts "at 1"
It would be equivalent to local  15..15 if the content for the next group of elements starts "at 2" or
23..23 if the next group of elements starts "at 3", etc.  I can see how that would be confusing.

With parameterized representation clauses, using the naming convention from the paper, you'd use MSB8 for the first case, MSB16 for the second, and MSB24 for the third, etc.  And yet it'd be nice to have a shorter, simpler way to write those clauses.  With no gaps, there's only one way to pack such records all together, so I mistakenly thought the new 'Bit_Order was looking after that, and that it was fixed more universally than it apparently has been.

**Comment by jere-software** [comment]
>Moreover, for `#3`, I'd guess that most types wouldn't even have representation clauses.

In my own case, there were quite a few.  I'm working to emulate an 8 bit system with a 16bit address space and a 24-32bit external address space.  There are tons of types that I would like to specify as LE, bus addresses are just a one off the type of my head.  I know the AI in question refers to record representation clauses, but even if we were just able to specify the byte order for scalar types, that would be a step up.  Is there room for a more constrained/well defined approach to applying byte order (forbid scenarios you feel are impossible, allow scenarios that are well defined)?
